rust/lance/src/dataset/scanner.rs

```diff
@@ -1765,6 +1782,80 @@ mod test {
         assert_eq!(knn.schema().field_names(), [DIST_COL, ROW_ID]);
     }
 
+    #[tokio::test]
+    async fn test_ann_prefilter() {
+        let test_dir = tempdir().unwrap();
+        let test_uri = test_dir.path().to_str().unwrap();
+
+        let schema = Arc::new(ArrowSchema::new(vec![
+            ArrowField::new("filterable", DataType::Int32, true),
+            ArrowField::new("vector", fixed_size_list_type(2, DataType::Float32), true),
+        ]));
+
+        let vector_values = Float32Array::from_iter_values((0..600).map(|x| x as f32));
+
+        let batches = vec![RecordBatch::try_new(
+            schema.clone(),
+            vec![
+                Arc::new(Int32Array::from_iter_values(0..300)),
+                Arc::new(FixedSizeListArray::try_new_from_values(vector_values, 2).unwrap()),
+            ],
+        )
+        .unwrap()];
+
+        let write_params = WriteParams::default();
+        let batches = RecordBatchIterator::new(batches.into_iter().map(Ok), schema.clone());
+        let mut dataset = Dataset::write(batches, test_uri, Some(write_params))
+            .await
+            .unwrap();
+
+        dataset
+            .create_index(
+                &["vector"],
+                IndexType::Vector,
+                None,
+                &VectorIndexParams::ivf_pq(2, 8, 2, false, MetricType::L2, 2),
+                false,
+            )
+            .await
+            .unwrap();
+
+        let query_key = Arc::new(Float32Array::from_iter_values((0..2).map(|x| x as f32)));
+        let mut scan = dataset.scan();
+        scan.filter("filterable > 5").unwrap();
+        scan.nearest("vector", &query_key, 1).unwrap();
+        scan.with_row_id();
+
+        let batches = scan
+            .try_into_stream()
+            .await
+            .unwrap()
+            .try_collect::<Vec<_>>()
+            .await
+            .unwrap();
+        assert_eq!(batches.len(), 0);
+
+        scan.prefilter(true);
+        let batches = scan
+            .try_into_stream()
+            .await
+            .unwrap()
+            .try_collect::<Vec<_>>()
+            .await
+            .unwrap();
+        assert_eq!(batches.len(), 1);
+
+        let first_match = batches[0]
+            .column_by_name(ROW_ID)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .unwrap()
+            .values()[0];
```

Comment:
FYI, if you don't need to handle the errors, there are some shorthands you can use here:
```suggestion
        let first_match = batches[0][ROW_ID].as_primitive::<UInt64type>()
            .values()[0];
```

Reply:
Thanks.  Switched to these.